File: 1123-lowest-common-ancestor-of-deepest-leaves/1123-lowest-common-ancestor-of-deepest-leaves.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public Pair<Integer, TreeNode> getLcaDeepestLeaves(TreeNode root) {
        if(root == null) {
            return new Pair<>(0, null);
        }
        Pair<Integer, TreeNode> left = getLcaDeepestLeaves(root.left);
        Pair<Integer, TreeNode> right = getLcaDeepestLeaves(root.right);
        if(left.getKey() == right.getKey()) {
            return new Pair<>(left.getKey() + 1, root);
        }
        else if(left.getKey() < right.getKey()) {
            return new Pair<>(right.getKey() + 1, right.getValue());
        }
        else {
            return new Pair<>(left.getKey() + 1, left.getValue());
        }
    }
    public TreeNode lcaDeepestLeaves(TreeNode root) {
        Pair<Integer, TreeNode> res = getLcaDeepestLeaves(root);
        return res.getValue();
    }
}

File: 1123-lowest-common-ancestor-of-deepest-leaves/1123-lowest-common-ancestor-of-deepest-leaves.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def lcaDeepestLeaves(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        def dfs(root):
            if root is None:
                return 0, None
            l_dep, l_lca = dfs(root.left)
            r_dep, r_lca = dfs(root.right)
            if l_dep == r_dep:
                return l_dep + 1, root
            
            if l_dep > r_dep:
                return l_dep + 1, l_lca
            else:
                return r_dep + 1, r_lca
        max_dep, lca = dfs(root)
        return lca